src/mainloop.rkt

```diff
@@ -192,307 +167,101 @@
+     [(alt prog '(patch) _) (program-body prog)]
+     [(alt _ _ (list)) #f]
+     [(alt _ _ (list prev)) (get-starting-expr prev)]))
 
-;; takes a subexpression alt and converts it to a full alt
-(define (reconstruct-alt altn orig)
-  (define-values (altn* loc-prefix)
+  ;; takes a patch and converts it to a full alt
+  (define (reconstruct-alt altn loc0 orig)
     (let loop ([altn altn])
       (match-define (alt prog event prev) altn)
       (cond
-       [(null? prev) (values altn '())]
-       [(equal? (car event) 'mainloop) (values (car prev) (cadr event))]
+       [(equal? event '(patch)) orig]
        [else
-        (define-values (altn* loc-prefix) (loop (car prev)))
-        (cond
-         [(null? loc-prefix) (values altn '())]
-         [else
-          (define evt*
-            (match event
-             [(list 'taylor name var loc)
-              (list 'taylor name var (append loc-prefix (cdr loc)))]
-             [(list 'change cng)
-              (match-define (change rule loc binds) cng)
-              (list 'change (change rule (append loc-prefix (cdr loc)) binds))]
-             [(list 'simplify loc)
-              (list 'simplify (append loc-prefix (cdr loc)))]))
-          (define prog* (location-do loc-prefix (alt-program orig) (λ (_) (program-body prog))))
-          (values (alt prog* evt* (list altn*)) loc-prefix)])])))
-  ; add to improve cache
-  (unless (null? loc-prefix)
-    (define orig-sub (location-get loc-prefix (alt-program orig)))
-    (cache-improvement! orig-sub altn))
-  altn*)
-
-;; cached alts have outdated location data
-(define (repair-cached-alt altn loc base)
-  (let loop ([altn altn])
-    (match altn
-     [(alt _ _ (list))
-      altn]
-     [(alt prog (list 'mainloop _) (list prev))
-      (alt prog (list 'mainloop loc) (list base))]
-     [(alt prog event (list prev))
-      (alt prog event (list (loop prev)))])))
-
-;; Converts subexpressions back to full alternatives
-(define (reconstruct!)
-  (define new-alts (map (curryr reconstruct-alt (^next-alt^)) (^gened-simplify^)))
-  (define prev-alts
-    (if (^duplocs^)
-        (apply append
-          (for/list ([(loc expr) (in-dict (^duplocs^))])
-            (for/list ([bad (in-list (hash-ref improve-cache expr))])
-              (define fixed (repair-cached-alt bad loc (^next-alt^)))
-              (reconstruct-alt fixed (^next-alt^)))))
-        (list)))
-  (^gened-full^ (append new-alts prev-alts)))
+        (define event*
+          (match event
+           [(list 'taylor name var loc)
+            (list 'taylor name var (append loc0 (cdr loc)))]
+           [(list 'change cng)
+            (match-define (change rule loc binds) cng)
+            (list 'change (change rule (append loc0 (cdr loc)) binds))]
+           [(list 'simplify loc)
+            (list 'simplify (append loc0 (cdr loc)))]))
+        (define prog* (location-do loc0 (alt-program orig) (λ (_) (program-body prog))))
+        (alt prog* event* (list (loop (first prev))))])))
```

Comment:
This is kind of ugly but I think for now it's probably the right move. Longer-term, we probably want to change how the `alt` structure works so we don't need to do this.